src/backend/x11/connection.rs

```diff
@@ -0,0 +1,135 @@
+};
+use x11rb::{rust_connection::ConnectError, xcb_ffi::XCBConnection};
+
+/// An error that may occur when connecting to the X server.
+#[derive(Debug, thiserror::Error)]
+pub enum ConnectToXError {
+    /// An error occured while setting up XCB.
+    #[error("XCB failed to connect to the X server")]
+    Xcb(ConnectError),
+
+    /// Some required libraries were not present.
+    #[error("Required libraries (xlib and xlib_libxcb) were not loaded")]
+    LibrariesNotLoaded,
+
+    /// Xlib failed to initialize it's connection.
+    #[error("An xlib connection could not be established")]
+    XlibNoConnection,
+
+    /// The Xlib implementation is not xcb based.
+    #[error("Xlib implementation does not support getting the inner xcb connection")]
+    NoXlibXcb,
+}
+
+impl From<ConnectError> for ConnectToXError {
+    fn from(inner: ConnectError) -> Self {
+        Self::Xcb(inner)
+    }
+}
+
+/// A connection to the X server.
+///
+/// This contains a way to access both the xcb connection and the xlib Display.
+pub struct XConnection {
+    xlib_library: Xlib,
+    /// The xlib Display that initiated the XCBConnection
+    ///
+    /// If we want to allow creation of an EGL Context using X11, the XCB extensions are not
+    /// implemented most drivers yet. So we also expose the Xlib types so an OpenGL context
+    /// can be made in nearly every driver.
+    xlib_display: *mut x11_dl::xlib::Display,
+    xcb_connection: XCBConnection,
+}
+
+impl XConnection {
+    /// Attempts to connect to the X server.
+    pub(crate) fn new(_logger: &Logger) -> Result<(XConnection, usize), ConnectToXError> {
+        // TODO: Log setup process
+
+        let xlib = Xlib::open().expect("Failed to open xlib library");
+        let xlib_xcb = Xlib_xcb::open().expect("Failed to load xlib_libxcb");
+        let (display, screen_number) = unsafe {
+            let display = (xlib.XOpenDisplay)(ptr::null());
+
+            if display.is_null() {
+                return Err(ConnectToXError::XlibNoConnection);
+            }
+
+            (display, (xlib.XDefaultScreen)(display))
+        };
+
+        // Transfer ownership of the event queue to XCB since we use XCB to handle events.
+        let xcb_connection_t = unsafe {
+            let ptr = (xlib_xcb.XGetXCBConnection)(display);
+            (xlib_xcb.XSetEventQueueOwner)(display, XEventQueueOwner::XCBOwnsEventQueue);
+            ptr
+        };
+
+        let xcb_connection = unsafe {
+            // Xlib implementation does not use XCB
+            if xcb_connection_t.is_null() {
+                (xlib.XCloseDisplay)(display);
+                return Err(ConnectToXError::NoXlibXcb);
+            }
+
+            // Do not drop the connection upon closure since Xlib created the xcb_connection_t.
+            // The Drop impl of `XConnection` will shutdown the Xlib Display
+            XCBConnection::from_raw_xcb_connection(xcb_connection_t, false)
+        };
+
+        if xcb_connection.is_err() {
```

Comment:
Off topic: I was briefly thinking "how can this be an error?!?". Then I looked at the code and saw that x11rb has to parse the `.setup()` that the X11 server sent here and that can go wrong (e.g. the server sent a too short setup claiming to contain 10000 screens). libxcb just does not handle this possibility at all...

Now I am happy about all the safety guarantees that Rust provides.

No one will get this, but since you already wrote on this issue: CC @DemiMarie x11rb is the Rust library I was talking about elsewhere (in case you even care about this - if not, that's totally fine as well).

Reply:
I do care!  For one, this means that Smithay has a chance to have XWayland support that is robust against malicious X servers, assuming that the X server someday gets sandboxed.  That would allow for per-Flatpak X servers that run in the Flatpak sandbox, for example.

Does x11rb use libxcb internally and then validate its output, or does it do everything itself?

Reply:
x11rb I believe decodes the events and encodes requests itself, libxcb only being used to read and write across the socket.

Reply:
> Does x11rb use libxcb internally and then validate its output, or does it do everything itself?

There are two options in x11rb. There is `RustConnection` that is pure rust. If you disable the `allow-unsafe-code` feature, then you lose support for FD-passing, but get all of Rust's safety guarantees always. 

With the `allow-unsafe-code` feature, there is also a `XCBConnection`. Just as @i509VCB, this only uses libxcb to send & receive raw bytes (This uses C APIs like `xcb_send_request` and `xcb_wait_for_event` / `xcb_wait_for_reply` from libxcb). Since libxcb does not provide an extra length return, the code has to look at something like three bytes of an X11 packet and figure out the length of the packet, then call the unsafe function `std::slice::from_raw_parts`. Afterwards, the same safe code is used as with `RustConnection`. I am pretty sure that these length calculations are exactly the same math as libxcb does internally, but obviously this is a bit less safe than `RustConnection`.

`XCBConnection` allows sharing the connection with C code that expects a libxcb `xcb_connection_t*` (e.g. cairo or mesa). `RustConnection` is a Rust-only thing that cannot be used in this way.

I bet no one cares, but x11rb even works on Windows. Obviously, there is no FD passing there.